### PR TITLE
Improve logging and add more messages

### PR DESCRIPTION
**Pull request by victorhurdugaci**
Fixes https://github.com/aspnet/StaticFiles/issues/69

please review @Tratcher @rynowak

### CODE REVIEW

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 1]
```diff
@@ -68,11 +68,38 @@ public StaticFileMiddleware(RequestDelegate next, IHostingEnvironment hostingEnv
...
-            if (fileContext.ValidateMethod()
-                && fileContext.ValidatePath()
-                && fileContext.LookupContentType()
-                && fileContext.LookupFileInfo())
+          
+            if (!fileContext.ValidateMethod())
             {
+                if (_logger.IsEnabled(LogLevel.Verbose))
+                {
+                    _logger.LogVerbose(LoggingEventIds.MethodMismatch, $"{context.Request.Method} requests are not supported");
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: Debug

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 2]
```diff
@@ -87,25 +114,31 @@ public Task Invoke(HttpContext context)
                         {
                             return fileContext.SendRangeAsync();
                         }
-                        if (_logger.IsEnabled(LogLevel.Verbose))
+                        if (_logger.IsEnabled(LogLevel.Debug))
                         {
-                            _logger.LogVerbose(string.Format("Copying file {0} to the response body", fileContext.SubPath));
+                            _logger.LogDebug(LoggingEventIds.CopyingFileToBody, string.Format("Copying file {0} to the response body", fileContext.SubPath));
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: Redundant

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 3]
```diff
@@ -87,25 +114,31 @@ public Task Invoke(HttpContext context)
...
                         }
+
+                        _logger.LogInformation(LoggingEventIds.FileServed, $"Sending file {fileContext.SubPath}");
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: Also include physical path if not empty?

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 4]
```diff
@@ -87,25 +114,31 @@ public Task Invoke(HttpContext context)
...
+                        if (_logger.IsEnabled(LogLevel.Verbose))
+                        {
+                            _logger.LogVerbose(LoggingEventIds.PreconditionFailed, string.Format("{0} precondition failed", fileContext.SubPath));
+                        }
                         return fileContext.SendStatusAsync(Constants.Status412PreconditionFailed);
 
                     default:
                         var exception = new NotImplementedException(fileContext.GetPreconditionState().ToString());
-                        _logger.LogError("No precondition state specified", exception);
+                        _logger.LogError(LoggingEventIds.NoPreconditionState, "No precondition state specified", exception);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: @Eilon We decided not to log exceptions that are about to be thrown, correct?

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 5]
```diff
@@ -68,11 +68,38 @@ public StaticFileMiddleware(RequestDelegate next, IHostingEnvironment hostingEnv
...
+                if (_logger.IsEnabled(LogLevel.Verbose))
+                {
+                    _logger.LogVerbose(LoggingEventIds.MethodMismatch, $"{context.Request.Method} requests are not supported");
+                }
+            }
+            else if (!fileContext.ValidatePath())
+            {
+                if (_logger.IsEnabled(LogLevel.Verbose))
+                {
+                    _logger.LogVerbose(LoggingEventIds.PathMismatch, $"The path {fileContext.SubPath} doesn't match the path filter");
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: Debug

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 6]
```diff
@@ -68,11 +68,38 @@ public StaticFileMiddleware(RequestDelegate next, IHostingEnvironment hostingEnv
...
+                }
+            }
+            else if (!fileContext.LookupContentType())
+            {
+                if (_logger.IsEnabled(LogLevel.Verbose))
+                {
+                    _logger.LogVerbose(LoggingEventIds.ContentMismatch, $"The path {fileContext.SubPath} is not for a supported file type");
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: This fires for requests even without extensions, correct? Debug.

**Comment by rynowak** [comment]
Take a look at da731fc641f56529328d3cb74d6092310a40627e - there's a super efficient and more slightly more cleanly pattern added recently

**Comment by victorhurdugaci** [comment]
Updated

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+        private static Action<ILogger, string, Exception> _logRangeNotSatisfiable;
+        private static Action<ILogger, StringValues, string, Exception> _logSendingFileRange;
+        private static Action<ILogger, StringValues, string, Exception> _logCopyingFileRange;
+        private static Action<ILogger, long, string, string, Exception> _logCopyingBytesToResponse;
+        private static Action<ILogger, string, Exception> _logMultipleFileRanges;
+
+        static LoggerExtensions()
+        {
+            _logMethodNotSupported = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Debug,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: I talked to Ryan and he said that most of these should be Verbose, so you always end up with at least one verbose message when the middleware does not handle a request. If it does handle the request then there should be one or more Informational messages.

File: src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs [hunk 8]
```diff
@@ -87,25 +102,21 @@ public Task Invoke(HttpContext context)
...
                         return fileContext.SendStatusAsync(Constants.Status304NotModified);
 
                     case StaticFileContext.PreconditionState.PreconditionFailed:
+                        _logger.LogPreconditionFailed(fileContext.SubPath);
                         return fileContext.SendStatusAsync(Constants.Status412PreconditionFailed);
 
                     default:
                         var exception = new NotImplementedException(fileContext.GetPreconditionState().ToString());
-                        _logger.LogError("No precondition state specified", exception);
+                        _logger.LogNoPrecondition(exception);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/StaticFileMiddleware.cs]: This should just be a Debug.Assert, not a log.

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+            _logFileNotFound = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
+                eventId: 5,
+                formatString: "The request path {Path} does not match an existing file");
+            _logPathNotModified = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
+                eventId: 5,
+                formatString: "The file {Path} was not modified");
+            _logPreconditionFailed = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: Informational, it's handling the request.

**Comment by victorhurdugaci** [comment]
Updated

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+            _logFileTypeNotSupported = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
+                eventId: 4,
+                formatString: "The request path {Path} does not match a supported file type");
+            _logFileNotFound = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
+                eventId: 5,
+                formatString: "The request path {Path} does not match an existing file");
+            _logPathNotModified = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: Info

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+                eventId: 5,
+                formatString: "The file {Path} was not modified");
+            _logPreconditionFailed = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Information,
+                eventId: 6,
+                formatString: "Precondition for {Path} failed");
+            _logHandled = LoggerMessage.Define<int, string>(
+                logLevel: LogLevel.Verbose,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: Info?

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+                eventId: 8,
+                formatString: "Handled. Status code: {StatusCode} File: {Path}");
+            _logRangeNotSatisfiable = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Verbose,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: Info or warn

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+                eventId: 9,
+                formatString: "Range not satisfiable for {Path}");
+            _logSendingFileRange = LoggerMessage.Define<StringValues, string>(
+                logLevel: LogLevel.Verbose,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: Info?

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+            _logCopyingFileRange = LoggerMessage.Define<StringValues, string>(
+                logLevel: LogLevel.Verbose,
+                eventId: 11,
+                formatString: "Copying {Range} of file {Path} to the response body");
+            _logCopyingBytesToResponse = LoggerMessage.Define<long, string, string>(
+                logLevel: LogLevel.Verbose,
+                eventId: 12,
+                formatString: "Copying bytes {Start}-{End} of file {Path} to response body");
+            _logMultipleFileRanges = LoggerMessage.Define<string>(
+                logLevel: LogLevel.Error,
```
Comment (Tratcher) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: warn, we'll still serve the file without ranges.

File: src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs (hunk 11 above)
Comment (victorhurdugaci) [inline src/Microsoft.AspNet.StaticFiles/LoggerExtensions.cs]: No. This is more of an implementation detail

**Comment by victorhurdugaci** [comment]
Updated again

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**